### ISSUE TITLE
Support WebAuthn

### ISSUE THREAD

**Issue by ghost** [ContainerSSH/ContainerSSH#90]
As suggested by @vorburger we should support [WebAuthn](https://webauthn.io/) for authentication purposes.

**Comment by janoszen** [comment]
Early demo of web-based authentication: https://youtu.be/ifP0xUraH20
/cc @vorburger

**Comment by janoszen** [comment]
The most likely path of getting Webauthn to work is via #56 since ContainerSSH doesn't have a user database itself and cannot present the user with a way to register their credentials. Since servers like [Keycloak](https://www.keycloak.org/) already solve this problem it would be good to create a guide for users how to set that up.

**Comment by kfox1111** [comment]
keycloak supports webauthn.  So once there is oidc support, that would come along for free?

I heard Keycloak supports device flow now. So that could be interesting to add support to containerssh for. The idea is, containerssh would contact keycloak via device flow. it would get a url back for the user to go to and then display it, and maybe even display a qr code of the url. The ssh session is blocked waiting for auth.

The user can visit the url via their browser, or even on their phone and authenticate via Keycloak. The ssh session then unblocks and continues on authenticated.

**Comment by janoszen** [comment]
@kfox1111 you are correct!

The device flow is intended for use cases like this, but SSH, specifically, is not well suited for this flow. The reason is that we print the link for the user along with the code (already hard to interpret in the console), then we would need the user to hit enter on a meaningless prompt once the authentication is done. We have no way to continue the authentication flow without interaction. Furthermore, providers like GitHub limit how many device flow authentications one can do per hour.

Instead, we are going with the normal authorization code flow. It works like this:

1. User enters username.
2. User client tries authentication methods, ends up using keyboard-interactive, gets presented with the link.
3. User clicks on link, gets redirected to the OAuth2 authorization code endpoint (Keycloak, GitHub, etc).
4. User logs in with whatever credentials they have.
5. User gets redirected back to a website served by ContainerSSH, which then prompts the user to copy a code (connection ID + authorization code) into the SSH console.
6. User copies the code into the console and hits enter
7. ContainerSSH fetches the access token from the provider.
8. ContainerSSH contacts the API of the provider (currently implementing it for OIDC and GitHub) to fetch the user information and verifies that the username entered into SSH matches the username on the remote provider.
9. ContainerSSH contacts the provider and deletes the access token if supported by the provider since it is no longer needed.

This sounds way more complicated than it really is, you can see it in action in this video: https://www.youtube.com/watch?v=ifP0xUraH20

We have contacted several SSH client vendors to improve compatibility and got back responses from 3, promising to improve compatibility. Regardless, this functionality should work as intended in OpenSSH, PuTTY and Filezilla. WinSCP and Bitvise have a patch coming to enable this.

**Comment by janoszen** [comment]
On second thought, we might actually be able to wait if we send a keyboard-interactive auth request with no questions and then wait, but we'll have to experiment how clients react to that.

**Comment by janoszen** [comment]
@kfox1111 looks like I have to revise my statement. You CAN use the device flow and it's even better than the other one. However, it won't work with GitHub due to their limit of 50 device auths per hour. See the demo here: https://youtu.be/SGHee9cV_rA

**Comment by janoszen** [comment]
Update on the QR code: it works on PuTTY, but OpenSSH cuts off the instruction after ~255 characters and also does not seem to support displaying UTF-8 characters. 
![](https://user-images.githubusercontent.com/662664/114544624-2e704a00-9c5b-11eb-8d2f-7fed9b935b7b.png)

**Comment by janoszen** [comment]
Published a devlog on the issue: https://containerssh.io/blog/2021/04/13/devlog-oauth2/

**Comment by kfox1111** [comment]
keycloak is getting much easier to deploy now that it has the keycloak-operator. Maybe that would ease the pain of deviceflow with github?

For the openssh instruction size, could you do multiple print commands to the client rather then try and do it all through one? (no idea if that would work)

**Comment by janoszen** [comment]
@kfox1111 we have a working device flow now. It's pretty janky, but we'll definitely support both. We can also do some client detection, so QR codes should be available for PuTTY. Patching OpenSSH is also on the table. We want to get this right so other SSH servers can follow in our path.

**Comment by kfox1111** [comment]
related... it might be interesting if you could pass the auth/id tokens into the ssh environment via variables. Could easily plumb that up to kubectl so that just sshing into containerssh would allow kubectl to work, no additional authentication needed as the actual user.

**Comment by janoszen** [comment]
That is also a good idea, haven't thought of that. 109 would be pretty close to that, but it'd be good if we had a mechanism to revoke the token when the connection finishes.

**Comment by janoszen** [comment]
@kfox1111 I just realized something: if we expose the GitHub tokens as an environment variable that would enable a user to automatically use the GitHub CLI after logging in, or potentially run `git clone` without needing to authenticate themselves.

**Comment by kfox1111** [comment]
That could be quite useful too. :)

Maybe a configurable though cause some users wouldn't like that.

**Comment by janoszen** [comment]
@kfox1111 yes indeed, that will be configurable.

**Comment by janosdebugs** [comment]
A little update after a username switch: the feature is currently well under way and here's a demonstration of a performed GitHub authentication and then the GitHub token being passed on to the container and used to query an avatar:

![image](https://user-images.githubusercontent.com/86970079/125000528-19c80600-e040-11eb-8fbf-d18fd6bfc004.png)

**Comment by tsipinakis** [comment]
This has been implemented as a feature-preview with 0.5. Closing
